src/swell/configuration/jedi/interfaces/geos_atmosphere/model/geometry_inner.py

```
@@ -0,0 +1,35 @@
+# (C) Copyright 2021- United States Government as represented by the Administrator of the
+# National Aeronautics and Space Administration. All Rights Reserved.
+#
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+
+# --------------------------------------------------------------------------------------------------
+
+from collections.abc import Mapping
+
+# --------------------------------------------------------------------------------------------------
+
+
+def geometry_inner(template_dict: Mapping) -> Mapping:
+
+    npx_proc = template_dict['npx_proc']
+    npy_proc = template_dict['npy_proc']
+    horizontal_resolution = template_dict['horizontal_resolution']
+    vertical_resolution = template_dict['vertical_resolution']
+
+    geometry_inner = {
+        'fms initialization': {
+            'namelist filename': './fv3-jedi/fv3files/fmsmpp.nml',
+            'field table filename': './fv3-jedi/fv3files/field_table_gmao'
+        },
+        'akbk': f'./fv3-jedi/fv3files/akbk{vertical_resolution}.nc4',
+        'layout': [npx_proc, npy_proc],
+        'npx': int(horizontal_resolution),
+        'npy': int(horizontal_resolution),
+        'npz': int(vertical_resolution)
```

This is good, and is a nice example of the advantages of having this live in Python code: We can enforce types early and loudly! (E.g., `int("3.0")` will fail loudly with a clear error).

That said, the "Parse, don't validate" philosophy (see suggested reading below) suggests that we should try to capture such type information as early as possible and carry it around with us. Here, that means that we should consider if we can define `template_dict` more precisely (e.g., as a dataclass with explicit types for all fields) so that all fields are validated and then all objects that work with `template_dict` know exactly what types everything should be.

Let's keep this for now. But flagging this as something to think about in future design.

https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/

---

On the other hand, just building validation directly into these functions is conceptually much simpler and maybe better for [Locality of Behavior](https://dev.to/ralphcone/new-hot-trend-locality-of-behavior-1g9k). So maybe it makes more sense to keep this type of structure. 🤷‍♂️ Food for thought!

---

In this case, the type enforcing was trying to recreate subtleties in the yaml formatting, as I was aiming to have there be zero diff between files generated by the old system and the new system. The value `horizontal_resolution` in the dictionary is a string, but when it was templated into yaml by the `jinja2` templating, it essentially becomes an int (It likely makes no difference, but I wasn't sure if whichever aspect of jedi uses this field would be able to handle a string vs an int). Your comment also made me realize that in some cases horizontal resolution does have to be a string.

---

OK, let's leave this as is for now to keep the zero diff. This is a bigger conversation.